---
title: "Trun"
output: pdf_document
---

**INTRODUCTION:**

This is the amplicon data analysis of my early SNF project at a hatchery in Trun, Switzerland. Briefly, I washed fish sperm with antibiotics and then used it to fertilize eggs. I am contrasting offspring of the same parents that originated from washed sperm with offspring that originated from 'natural' sperm.

```{r, echo=FALSE}
library(ggplot2)
library(ggmap)
library(maps)
library(mapdata)


bikemap2 <- get_map(location = c(8.991, 46.7432), maptype = "terrain", source = "google", zoom = 5)
gg1 <- ggmap(bikemap2)

labs <- data.frame(
  long = 8.991,
  lat = 46.7432,
  names = "Hatchery",
  stringsAsFactors = FALSE
)  

gg1 + 
  geom_point(data = labs, aes(x = long, y = lat), color = "red", size = 3)
```

First I needed to download dada2.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r eval=FALSE}
# Install dada2 through bioconductor

source("http://bioconductor.org/biocLite.R")
biocLite("dada2")
```


I ran most of this on Oskar. Oskar is a powerful Mac in the Eisen lab.
I used the following commands:

ssh megaptera@128.120.143.102

scp -r /Users/megaptera/Desktop/shared_helper/dada/silva_nr_v128_train_set.fa megaptera@128.120.143.102:/Users/megaptera/dada2

scp -r megaptera@128.120.143.102:/Users/megaptera/dada2/outputs /Users/megaptera/Desktop

--> To use R on Oskar:

R # simply type R.

I followed this Stanford tutorial by Benjamin Callahan:
http://web.stanford.edu/class/bios221/MicrobiomeWorkflowII.html


**FILE PARSING:**

I did two rounds of sequencing so I will do QC twice individually before merging the files. 

My samples were sequenced at the IMR in Dalhousie. Some samples (mostly sperm) had very low DNA concentrations to start with. There were only very few resulting sequencing reads using the standard Dalhousie pipeline. I asked them to re-run my samples after verifying that the PCR resulted in nice bands. They did that for me. 

--> For the second round of sequencing there were more PCR amplification rounds.

```{r eval=FALSE}
library(dada2); packageVersion("dada2")
# File parsing for the first round of sequencing
pathF_round1  <- "/Users/megaptera/dada2/dada_firstRound/fastq/pathF" # CHANGE ME to the directory containing the fastq files before unzipping.
pathR_round1 <- "/Users/megaptera/dada2/dada_firstRound/fastq/pathR"
filtpathF_round1 <- file.path(pathF_round1, "filtered") # Filtered forward files go into the pathF/filtered/ subdirectory
filtpathR_round1 <- file.path(pathR_round1, "filtered") # ...
fastqFs_round1 <- sort(list.files(pathF_round1, pattern="fastq.gz"))
fastqRs_round1 <- sort(list.files(pathR_round1, pattern="fastq.gz"))
if(length(fastqFs_round1) != length(fastqRs_round1)) stop("Forward and reverse files do not match.")
# Filtering: THESE PARAMETERS ARENT OPTIMAL FOR ALL DATASETS; truncQ=2 for Guillaume.
filterAndTrim(fwd=file.path(pathF_round1, fastqFs_round1), filt=file.path(filtpathF_round1, fastqFs_round1),
              rev=file.path(pathR_round1, fastqRs_round1), filt.rev=file.path(filtpathR_round1, fastqRs_round1),
              truncLen=c(280,200), maxEE=c(2,5), maxN=0,
              compress=TRUE, verbose=TRUE, multithread=TRUE)


# File parsing for the second round of sequencing
pathF_round2  <- "/Users/megaptera/dada2/dada_secondRound/fastq/pathF" # CHANGE ME to the directory containing the fastq files before unzipping.
pathR_round2 <- "/Users/megaptera/dada2/dada_secondRound/fastq/pathR"
filtpathF_round2 <- file.path(pathF_round2, "filtered") # Filtered forward files go into the pathF/filtered/ subdirectory
filtpathR_round2 <- file.path(pathR_round2, "filtered") # ...
fastqFs_round2 <- sort(list.files(pathF_round2, pattern="fastq.gz"))
fastqRs_round2 <- sort(list.files(pathR_round2, pattern="fastq.gz"))
if(length(fastqFs_round2) != length(fastqRs_round2)) stop("Forward and reverse files do not match.")
# Filtering: THESE PARAMETERS ARENT OPTIMAL FOR ALL DATASETS
filterAndTrim(fwd=file.path(pathF_round2, fastqFs_round2), filt=file.path(filtpathF_round2, fastqFs_round2),
              rev=file.path(pathR_round2, fastqRs_round2), filt.rev=file.path(filtpathR_round2, fastqRs_round2),
              truncLen=c(280,200), maxEE=c(2,5), maxN=0,
              compress=TRUE, verbose=TRUE, multithread=TRUE)
```



**INFERRING SEQUENCING VARIANTS:**

I ran this in interactive mode and not with a script. Be aware that I used the same variable names for the two different sequencing runs. This has to be run separately. Otherwise you will overwrite the first variables with the second ones! 

```{r eval=FALSE}
filtpathF_round1_filt <- "/Users/megaptera/dada2/dada_firstRound/fastq/pathF/filtered"
filtpathR_round1_filt <- "/Users/megaptera/dada2/dada_firstRound/fastq/pathR/filtered"
filtFs_round1_filt <- list.files(filtpathF_round1_filt, pattern="fastq.gz", full.names = TRUE)
filtRs_round1_filt <- list.files(filtpathR_round1_filt, pattern="fastq.gz", full.names = TRUE)
sample.names_round1_filt <- sapply(strsplit(basename(filtFs_round1_filt), "_"), `[`, 1) # Assumes filename = samplename_XXX.fastq.gz
sample.namesR_round1_filt <- sapply(strsplit(basename(filtRs_round1_filt), "_"), `[`, 1) # Assumes filename = samplename_XXX.fastq.gz
if(!identical(sample.names_round1_filt, sample.namesR_round1_filt)) stop("Forward and reverse files do not match.")
names(filtFs_round1_filt) <- sample.names
names(filtRs_round1_filt) <- sample.names
set.seed(20010318)
# Learn forward error rates
errF_round1_filt <- learnErrors(filtFs_round1_filt, nread=1e6, multithread=TRUE)
# Learn reverse error rates
errR_round1_filt <- learnErrors(filtRs_round1_filt, nread=1e6, multithread=TRUE)
# Sample inference and merger of paired-end reads
# plotErrors(errF, nominalQ=TRUE)
mergers <- vector("list", length(sample.names_round1_filt))
names(mergers_round1_filt) <- sample.names_round1_filt
for(sam in sample.names_round1_filt) {
  cat("Processing:", sam, "\n")
  derepF_round1_filt <- derepFastq(filtFs_round1_filt[[sam]])
  ddF_round1_filt <- dada(derepF_round1_filt, err=errF_round1_filt, multithread=TRUE)
  derepR_round1_filt <- derepFastq(filtRs_round1_filt[[sam]])
  ddR_round1_filt <- dada(derepR_round1_filt, err=errR_round1_filt, multithread=TRUE)
  merger_round1_filt <- mergePairs(ddF_round1_filt, derepF_round1_filt, ddR_round1_filt, derepR_round1_filt)
  mergers[[sam]] <- merger_round1_filt
}
rm(derepF_round1_filt); rm(derepR_round1_filt)
# Construct sequence table and remove chimeras
seqtab_round1_filt <- makeSequenceTable(mergers_round1_filt)
write.table(seqtab_round1_filt, file="test_seqtab_sequencing_firstRound_March18.txt")
saveRDS(seqtab_round1_filt, "/Users/megaptera/dada2/outputs/seqtab_firstRound_March18.rds") # CHANGE ME to where you want sequence table saved


filtpathF <- "/Users/megaptera/dada2/dada_secondRound/fastq/pathF/filtered"
filtpathR <- "/Users/megaptera/dada2/dada_secondRound/fastq/pathR/filtered"
filtFs <- list.files(filtpathF, pattern="fastq.gz", full.names = TRUE)
filtRs <- list.files(filtpathR, pattern="fastq.gz", full.names = TRUE)
sample.names <- sapply(strsplit(basename(filtFs), "_"), `[`, 1) # Assumes filename = samplename_XXX.fastq.gz
sample.namesR <- sapply(strsplit(basename(filtRs), "_"), `[`, 1) # Assumes filename = samplename_XXX.fastq.gz
if(!identical(sample.names, sample.namesR)) stop("Forward and reverse files do not match.")
names(filtFs) <- sample.names
names(filtRs) <- sample.names
set.seed(1001)
# Learn forward error rates
errF <- learnErrors(filtFs, nread=1e6, multithread=TRUE)
# Learn reverse error rates
errR <- learnErrors(filtRs, nread=1e6, multithread=TRUE)
# Sample inference and merger of paired-end reads
# plotErrors(errF, nominalQ=TRUE)
mergers <- vector("list", length(sample.names))
names(mergers) <- sample.names
for(sam in sample.names) {
  cat("Processing:", sam, "\n")
  derepF <- derepFastq(filtFs[[sam]])
  ddF <- dada(derepF, err=errF, multithread=TRUE)
  derepR <- derepFastq(filtRs[[sam]])
  ddR <- dada(derepR, err=errR, multithread=TRUE)
  merger <- mergePairs(ddF, derepF, ddR, derepR)
  mergers[[sam]] <- merger
}
rm(derepF); rm(derepR)
# Construct sequence table and remove chimeras
seqtab2 <- makeSequenceTable(mergers)
write.table(seqtab2, file="test_seqtab_sequencing_secondRound_Feb18.txt")
saveRDS(seqtab2, "/Users/megaptera/dada2/outputs/seqtab_secondRound_Feb18.rds") # CHANGE ME to where you want sequence table saved

biocLite("dada2")
```

**MERGING, REMOVING CHIMERAS AND ASSIGNING TAXONOMY CAN BE DONE TOGETHER AGAIN!**

```{r eval=FALSE}
# Merge multiple runs (if necessary)
st1 <- readRDS("/Users/megaptera/dada2/outputs/seqtab_secondRound_Feb18.rds")
st2 <- readRDS("/Users/megaptera/dada2/outputs/seqtab_firstRound_Feb18.rds")
st.all <- mergeSequenceTables(st1, st2)

# Remove chimeras
seqtab.chim <- removeBimeraDenovo(st.all, method="consensus", multithread=TRUE)

# send newest silva training set to Oskar!
scp /Users/megaptera/Desktop/silva_nr_v132_train_set.fa megaptera@128.120.143.102:/Users/megaptera/dada2

# Assign taxonomy
tax <- assignTaxonomy(seqtab.chim, "/Users/megaptera/dada2/silva_nr_v132_train_set.fa", multithread=TRUE)
#taxa <- addSpecies(taxa, "Training/silva_species_assignment_v128.fa.gz")


# Write to disk
saveRDS(seqtab.chim, "/Users/megaptera/dada2/outputs/seqtab_final_chim_Feb18.rds") # CHANGE ME to where you want sequence table saved
saveRDS(tax, "/Users/megaptera/dada2/outputs/tax_final_Feb18.rds") # CHANGE ME ...
```

**NOW HAND IT OVER TO PHYLOSEQ!**

First you need to read in the data again!

```{r}
setwd("/Users/megaptera/Desktop/Finish_that_Trun_analysis/working_directory/")

taxa <- readRDS("tax_final_Feb18.rds")
seqtab.nochim <- readRDS("seqtab_final_chim_Feb18.rds")

write.table(taxa, file="taxa_table.txt")
write.table(seqtab.nochim, file="seqtab_table.txt")

samdf <- read.delim("phyloseq_map.txt", header=T)
samdf <- as.data.frame(samdf)
row.names(samdf) <- samdf$Subject


source("https://bioconductor.org/biocLite.R")
biocLite("phyloseq")

library(phyloseq); packageVersion("phyloseq")
library(ggplot2); packageVersion("ggplot2")
```

Here is a list of R environment based tools for 16S rRNA gene data exploration, statistical analysis and visualization
# https://microsud.github.io/Tools-Microbiome-Anlaysis/

**How to get your tree for phylogenetic approaches**

If you want to do phylogenetic approaches you need a tree!
This is what you do to get a tree:

Phylogenetic relatedness is commonly used to inform downstream analyses, especially the calculation of phylogeny-aware distances between microbial communities. The DADA2 sequence inference method is reference-free, so we must construct the phylogenetic tree relating the inferred sequence variants de novo. We begin by performing a multiple-alignment using the DECIPHER R package (Wright 2015).

```{r eval=FALSE}
# First we need to get the right libraries.
library("knitr")
#source("https://bioconductor.org/biocLite.R")
#biocLite("BiocStyle")
library("BiocStyle")
.cran_packages <- c("ggplot2", "gridExtra")
.bioc_packages <- c("dada2", "phyloseq", "DECIPHER", "phangorn")
.inst <- .cran_packages %in% installed.packages()
if(any(!.inst)) {
  install
  
  
  .packages(.cran_packages[!.inst])
}
.inst <- .bioc_packages %in% installed.packages()
if(any(!.inst)) {
  source("http://bioconductor.org/biocLite.R")
  biocLite(.bioc_packages[!.inst], ask = F)
}
# Load packages into session, and print package version
sapply(c(.cran_packages, .bioc_packages), require, character.only = TRUE)


seqs <- getSequences(seqtab.nochim)
names(seqs) <- seqs # This propagates to the tip labels of the tree
alignment <- AlignSeqs(DNAStringSet(seqs), anchor=NA,verbose=FALSE)
```

The phangorn R package is then used to construct a phylogenetic tree. Here we first construct a neighbor-joining tree, and then fit a GTR+G+I (Generalized time-reversible with Gamma rate variation) maximum likelihood tree using the neighbor-joining tree as a starting point.

```{r eval=FALSE}
phangAlign <- phyDat(as(alignment, "matrix"), type="DNA")
dm <- dist.ml(phangAlign)
treeNJ <- NJ(dm) # Note, tip order != sequence order
fit = pml(treeNJ, data=phangAlign)
fitGTR <- update(fit, k=4, inv=0.2)
fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
                    rearrangement = "stochastic", control = pml.control(trace = 0))
detach("package:phangorn", unload=TRUE)
```

I ran this all on Oskar!

my script looked like this, and I ran it with the following command:

screen
R CMD BATCH /Users/megaptera/dada2/outputs/Oskar_tree.R
ctrl-a d
screen -r

If you want to understand ordination metrices better go to this tutorial:
http://albertsenlab.org/ampvis2-ordination/


```{r eval=FALSE}
setwd("/Users/megaptera/dada2/outputs/")

taxa <- readRDS("tax_final_Feb18.rds")
seqtab.nochim <- readRDS("seqtab_final_chim_Feb18.rds")

#rownames(taxa)
#rownames(seqtab.nochim)

#write.table(taxa, file="taxa_table_test_for_R.txt")
#write.table(seqtab.nochim, file="seqtab_table_test_for_R.txt")

#source("https://bioconductor.org/biocLite.R")
#biocLite("phyloseq")

library(phyloseq); packageVersion("phyloseq")
library(ggplot2); packageVersion("ggplot2")

# A list of R environment based tools for 16S rRNA gene data exploration, statistical analysis and visualization
# https://microsud.github.io/Tools-Microbiome-Anlaysis/


# If you want to do phylogenetic approaches you need a tree!
# This is what you do to get a tree:

# Phylogenetic relatedness is commonly used to inform downstream analyses, especially the calculation of phylogeny-aware distances between microbial communities. The DADA2 sequence inference method is reference-free, so we must construct the phylogenetic tree relating the inferred sequence variants de novo. We begin by performing a multiple-alignment using the DECIPHER R package (Wright 2015).

# First we need to get the right libraries.
library("knitr")
#source("https://bioconductor.org/biocLite.R")
#biocLite("BiocStyle")
library("BiocStyle")
.cran_packages <- c("ggplot2", "gridExtra")
.bioc_packages <- c("dada2", "phyloseq", "DECIPHER", "phangorn")
.inst <- .cran_packages %in% installed.packages()
if(any(!.inst)) {
  install.packages(.cran_packages[!.inst])
}
.inst <- .bioc_packages %in% installed.packages()
if(any(!.inst)) {
  source("http://bioconductor.org/biocLite.R")
  biocLite(.bioc_packages[!.inst], ask = F)
}
# Load packages into session, and print package version
sapply(c(.cran_packages, .bioc_packages), require, character.only = TRUE)


seqs <- getSequences(seqtab.nochim)
names(seqs) <- seqs # This propagates to the tip labels of the tree
alignment <- AlignSeqs(DNAStringSet(seqs), anchor=NA,verbose=FALSE)
saveRDS(alignment, "/Users/megaptera/dada2/outputs/alignment_Feb18.rds")



# The phangorn R package is then used to construct a phylogenetic tree. Here we first construct a neighbor-joining tree, and then fit a GTR+G+I (Generalized time-reversible with Gamma rate variation) maximum likelihood tree using the neighbor-joining tree as a starting point.
phangAlign <- phyDat(as(alignment, "matrix"), type="DNA")
dm <- dist.ml(phangAlign)
treeNJ <- NJ(dm) # Note, tip order != sequence order
fit = pml(treeNJ, data=phangAlign)
fitGTR <- update(fit, k=4, inv=0.2)
fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
                    rearrangement = "stochastic", control = pml.control(trace = 0))
detach("package:phangorn", unload=TRUE)
saveRDS(fitGTR, "/Users/megaptera/dada2/outputs/fitGTR_Feb18.rds")


library(ape)

write.tree(fitGTR$tree, file="my_first_Oskar_tree.tre")
```

**now lets make some phyloseq plots :**

```{r}
# Open the libraries:
library(phyloseq); packageVersion("phyloseq")
library(ggplot2); packageVersion("ggplot2")

# Read in the data:
setwd("/Users/megaptera/Desktop/Finish_that_Trun_analysis/working_directory/")

taxa <- readRDS("tax_final_Feb18.rds")
seqtab.nochim <- readRDS("seqtab_final_chim_Feb18.rds")
fitGTR <- readRDS("fitGTR_Feb18.rds")


samdf <- read.delim("phyloseq_map.txt", header=T)
samdf <- as.data.frame(samdf)
row.names(samdf) <- samdf$Subject

ps.w.tree <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               sample_data(samdf), 
               tax_table(taxa), phy_tree(fitGTR$tree))
```

**Here it has to be noted that my phangorn tree is unrooted!**
We might want to root it first!
  
```{r}
require("ape")
phyloseq.tree <- root(fitGTR$tree, outgroup = "GTGTCAGCCGCCGCGGTAAAACCAGCACCTCAAGTGGTCAGGATGATTATTGGGCCTAAAGCATCCGTAGCCTGTTCTGTAAGTTTTCGGTTAAATCCATAAGCTCAACTTATGGGCTGCCGAAAATACTGCAGGACTAGGGAGTGGGAGAGGTAGACGGTACTCGGTAGGAAGGGGTAAAATCCTTTGATCTATTGATGACCACCTGTGGCGAAGGCGGTCTACCAGAACACGTTCGACGGTGAGGGATGAAAGCTGGGGGAGCAAACCGGATTAGATACCCGGGTAGTCCCAGCTGTAAACGATGCAAACTCGATGATGCGTTGGCTTGTTGCTAGCGCAGTGTCGCAGGGAAGCCGTTAAGTTTGCCGCCTGGGAAGTACGTACGCAAGTATGAAACTGAAATAAATTGACGG", resolve.root = TRUE)
```

Now we can continue with phyloseq:

```{r}
ps.w.tree.root <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               sample_data(samdf), 
               tax_table(taxa), phy_tree(phyloseq.tree))

# Distance plots
ord.nmds.bray <- ordinate(ps.w.tree.root, method="NMDS", distance="bray")
plot_ordination(ps.w.tree.root, ord.nmds.bray, color="Sequencing", title="Bray NMDS")

ord.PCoA.UniFrac <- ordinate(ps.w.tree, method="PCoA", distance="unifrac")
plot_ordination(ps.w.tree, ord.PCoA.UniFrac, color="Sequencing", title="UniFrac PCoA")

ord.PCoA.UniFrac.w.tree.root <- ordinate(ps.w.tree.root, method="PCoA", distance="unifrac")
plot_ordination(ps.w.tree.root, ord.PCoA.UniFrac.w.tree.root, color="Sequencing", title="UniFrac PCoA rooted tree", label="Treatment")
plot_ordination(ps.w.tree.root, ord.PCoA.UniFrac.w.tree.root, color="Treatment", title="UniFrac PCoA rooted tree")
plot_ordination(ps.w.tree.root, ord.PCoA.UniFrac.w.tree.root, color="Population", title="UniFrac PCoA rooted tree")
plot_ordination(ps.w.tree.root, ord.PCoA.UniFrac.w.tree.root, color="Tissue2", title="UniFrac PCoA rooted tree")

```
Here you can see the two different sequencing runs in red and green!

Now only look at sperm:

```{r eval=FALSE, echo=FALSE}
ps <- subset_samples(ps.w.tree.root, SelectionMilt=="sperm")


# Richness plots
plot_richness(ps, x="FishID2", measures=c("Shannon", "Simpson"), color="Tissue") + theme_bw()

# Distance plots
ord.nmds.bray <- ordinate(ps, method="NMDS", distance="bray")
plot_ordination(ps, ord.nmds.bray, color="FishID2", title="Bray NMDS")

# Barplot
top20 <- names(sort(taxa_sums(ps), decreasing=TRUE))[1:20]
ps.top20 <- transform_sample_counts(ps, function(OTU) OTU/sum(OTU))
ps.top20 <- prune_taxa(top20, ps.top20)
plot_bar(ps.top20, x="FishID2", fill="Genus") + facet_wrap(~FishID2, scales="free_x")

# Heatmap
gpt <- subset_taxa(ps, Kingdom=="Bacteria")
gpt <- prune_taxa(names(sort(taxa_sums(gpt),TRUE)[1:300]), gpt)
plot_heatmap(gpt, sample.label="Tissue")
```

Let's only look at fertilized eggs:

```{r}
# Read in the data:
setwd("/Users/megaptera/Desktop/Finish_that_Trun_analysis/working_directory/")

taxa <- readRDS("tax_final_Feb18.rds")
seqtab.nochim <- readRDS("seqtab_final_chim_Feb18.rds")
fitGTR <- readRDS("fitGTR_Feb18.rds")


samdf <- read.delim("phyloseq_map.txt", header=T)
samdf <- as.data.frame(samdf)
row.names(samdf) <- samdf$Subject

ps.w.tree <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               sample_data(samdf), 
               tax_table(taxa), phy_tree(fitGTR$tree))

require("ape")
phyloseq.tree <- root(fitGTR$tree, outgroup = "GTGTCAGCCGCCGCGGTAAAACCAGCACCTCAAGTGGTCAGGATGATTATTGGGCCTAAAGCATCCGTAGCCTGTTCTGTAAGTTTTCGGTTAAATCCATAAGCTCAACTTATGGGCTGCCGAAAATACTGCAGGACTAGGGAGTGGGAGAGGTAGACGGTACTCGGTAGGAAGGGGTAAAATCCTTTGATCTATTGATGACCACCTGTGGCGAAGGCGGTCTACCAGAACACGTTCGACGGTGAGGGATGAAAGCTGGGGGAGCAAACCGGATTAGATACCCGGGTAGTCCCAGCTGTAAACGATGCAAACTCGATGATGCGTTGGCTTGTTGCTAGCGCAGTGTCGCAGGGAAGCCGTTAAGTTTGCCGCCTGGGAAGTACGTACGCAAGTATGAAACTGAAATAAATTGACGG", resolve.root = TRUE)

ps.w.tree.root <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               sample_data(samdf), 
               tax_table(taxa), phy_tree(phyloseq.tree))
               
ps <- subset_samples(ps.w.tree.root, TavanasaAndValsAPS=="yes")

plot_richness(ps, x="Population", measures=c("Shannon", "Simpson"), color="Treatment") + theme_bw()


plot_ordination(ps, PCoA.eggs, color="Treatment", title="UniFrac PCoA rooted tree", label="Sequencing")

gpt <- subset_taxa(ps, Kingdom=="Bacteria")
gpt <- prune_taxa(names(sort(taxa_sums(gpt),TRUE)[1:300]), gpt)
ord.nmds.bray <- ordinate(ps, method="NMDS", distance="bray")
plot_ordination(ps, ord.nmds.bray, color="Treatment", title="Bray NMDS")
ord.nmds.bray <- ordinate(ps, method="NMDS", distance="bray")
plot_ordination(ps, ord.nmds.bray, color="Sequencing", title="Bray NMDS")

PCoA.eggs <- ordinate(ps, method="PCoA", distance="unifrac")
plot_ordination(ps, PCoA.eggs, color="Sequencing", title="UniFrac PCoA rooted tree", label="Treatment")
plot_heatmap(gpt, sample.label="Treatment")

gpt <- subset_taxa(ps, Kingdom=="Bacteria")
gpt <- prune_taxa(names(sort(taxa_sums(gpt),TRUE)[1:300]), gpt)
plot_heatmap(gpt, sample.label="Population")

# Barplot
top20 <- names(sort(taxa_sums(ps), decreasing=TRUE))[1:20]
ps.top20 <- transform_sample_counts(ps, function(OTU) OTU/sum(OTU))
ps.top20 <- prune_taxa(top20, ps.top20)
plot_bar(ps.top20, x="Treatment", fill="Genus") + facet_wrap(~Population, scales="free_x")

Jliv <- subset_taxa(ps, Phylum=="Proteobacteria")
plot_bar(Jliv, x="Treatment", fill="Genus") + facet_wrap(~Population, scales="free_x")
```



**Much later I decided to look for J. liv in the milt data without filtering it first!**





**R Markdown**

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

**Including Plots**

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
